server/service/sessions.go

```diff
@@ -594,3 +595,82 @@ func (svc *Service) validateSession(ctx context.Context, session *fleet.Session)
 
 	return svc.ds.MarkSessionAccessed(ctx, session)
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// Demo Login
+////////////////////////////////////////////////////////////////////////////////
+
+// This is a special kind of login where the username and password come in form values rather than
+// JSON as is typical for API requests. This is intended to support logins from demo environments,
+// when users are being redirected from fleetdm.com.
+
+type demologinRequest struct {
+	Email    string
+	Password string
+}
+
+func (demologinRequest) DecodeRequest(ctx context.Context, r *http.Request) (interface{}, error) {
+	err := r.ParseForm()
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, &badRequestError{message: err.Error()}, "decode demo login")
+	}
+
+	return demologinRequest{
+		Email:    r.FormValue("email"),
+		Password: r.FormValue("password"),
+	}, nil
+}
+
+type demologinResponse struct {
+	content string
+	Err     error `json:"error,omitempty"`
+}
+
+func (r demologinResponse) error() error { return r.Err }
+
+// If html is present we return a web page
+func (r demologinResponse) html() string { return r.content }
+
+func makeDemologinEndpoint(urlPrefix string) handlerFunc {
+	return func(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+		if os.Getenv("FLEET_DEMO") != "1" {
+			return nil, errors.New("this endpoint only enabled in demo mode")
+		}
+		req := request.(demologinRequest)
+
+		_, sess, err := svc.Login(ctx, req.Email, req.Password)
+
+		session := struct {
+			Token string
+		}{}
+		var resp demologinResponse
+		if err != nil {
+			resp.Err = err
```

Comment:
If `svc.Login` fails, does it make sense to continue execution?

Reply:
EDIT: And looks like if `err == nil` then `sess != nil` (so no need to check the latter AFAICS).

Sorry for the confusion. 

This is what I meant to say:
```go
1. if err != nil {
2.  resp.Err = err
3.  return resp, nil
4. }
5. // use sess
```

If you don't want to return in line 3 when `err != nil` then yes, you need to check for `sess != nil`.